primitives/enclave-bridge/src/lib.rs

```diff
@@ -34,3 +34,60 @@ pub struct ShardSignerStatus<AccountId, BlockNumber> {
 	pub fingerprint: EnclaveFingerprint,
 	pub last_activity: BlockNumber,
 }
+
+#[derive(Encode, Decode, Clone, PartialEq, Eq, sp_core::RuntimeDebug, TypeInfo)]
+pub struct ShardConfig<AccountId> {
+	/// enclave fingerprint which may perform state transitions on this shard
+	pub enclave_fingerprint: EnclaveFingerprint,
+	/// an optional limit on the number of validateers
+	pub max_instances: Option<u32>,
+	/// an optional set of authorities for permissioned sidechains
+	pub authorities: Option<Vec<AccountId>>,
+	/// maintenance mode blocks any upcoming state transitions on this shard
+	pub maintenance_mode: bool,
+}
+
+impl<AccountId> ShardConfig<AccountId> {
+	pub fn new(fingerprint: EnclaveFingerprint) -> Self {
+		ShardConfig {
+			enclave_fingerprint: fingerprint,
+			max_instances: None,
+			authorities: None,
+			maintenance_mode: false,
+		}
+	}
+}
+
+#[derive(Encode, Decode, Clone, PartialEq, Eq, sp_core::RuntimeDebug, TypeInfo)]
+pub struct UpgradableShardConfig<AccountId, BlockNumber> {
```

Comment:
I see that this does simplify the code. The only argument I see for my approach is that it needs one less less db lookup. This may actually be relevant because every sidechain finalization will perform this

Reply:
In face of the elegance vs. performance tradeoff, pallet design should pick performance

Reply:
Hmm, I am uncertain if I agree with this:

* it is not about elegance in my opinion, but about separation of concerns, which will make our life easier in the future
* i think we might even come up with a design that we don't need to check this upon every finalization: we could schedule the upgrade at block x with the scheduler pallet, and then we only have to check the current configuration upon block finalization.

Reply:
SOLID principles need to yield to performance considerations in low-level designs - such as pallets IME. And I don't even think that "single responsability" is severly questionable here. there's only one reason for `UpgradableShardConfig` to change.

Scheduler should be avoided whenever possible, because it is brittle. You should never make assumptions on how the runtime looks like at any block in the future

Reply:
Well, I see problems here, which are not resolvable without that separation, but I will not stand in the way of this PR.

Reply:
experiment. simulate another storage map lookup:
```
diff --git a/enclave-bridge/src/lib.rs b/enclave-bridge/src/lib.rs
index 239aeb0..218844b 100644
--- a/enclave-bridge/src/lib.rs
+++ b/enclave-bridge/src/lib.rs
@@ -346,6 +346,9 @@ impl<T: Config> Pallet<T> {
                current_block_number: T::BlockNumber,
                apply_due_update: bool,
        ) -> Option<ShardConfig<T::AccountId>> {
+               if Self::shard_config(ShardIdentifier::from([43u8; 32])).is_some() {
+                       log::info!("dummy")
+               };
                Self::shard_config(shard).map(|current| {
                        current.upgrade_at.filter(|&at| at <= current_block_number).map_or_else(
                                || current.active_config.clone(),

```
yields a benchmark diff of:
```
        fn confirm_processed_parentchain_block() -> Weight {
                // Proof Size summary in bytes:
                //  Measured:  `350`
-               //  Estimated: `3815`
-               // Minimum execution time: 18_077_000 picoseconds.
-               Weight::from_parts(23_090_000, 0)
-                       .saturating_add(Weight::from_parts(0, 3815))
-                       .saturating_add(T::DbWeight::get().reads(3))
+               //  Estimated: `6290`
+               // Minimum execution time: 45_935_000 picoseconds.
+               Weight::from_parts(47_374_000, 0)
+                       .saturating_add(Weight::from_parts(0, 6290))
+                       .saturating_add(T::DbWeight::get().reads(4))
                        .saturating_add(T::DbWeight::get().writes(1))
        }
```
measured execution time is highly unreliable, but as expected:
* db reads goes up by one (and so does PoV size)
* Estimate goes up significantly
* Measurement goes up significantly

The extrinsic `confirm_processed_parentchain_block` is our number one bottleneck when it comes to the limit of number of sidechains on our network. So every ns counts for this rather lightweight extrinsic